### PR TITLE
Support for utf8 chars

### LINKED ISSUE #6: if `Buffer#toString` not consume all bytes?
If i write data one byte by one byte, it emit empty.
```
var split = require("split2");
var buf = new Buffer("烫烫烫\r\n锟斤拷", "utf8");
var s = split();
s.on("data", function (data) { 
	console.log(data);
});

for (var i = 0; i < buf.length; ++i)
{
	s.write(buf.slice(i, i + 1));
}
s.end();
```

### PR DESCRIPTION
**Pull request by mcollina**
Fixes #6.

However, __I am not happy with this solution__, because it depends on https://www.npmjs.com/package/utf-8-validate.

This is a highly popular module, so possibly we should look for a different solution that is JS-only, possibly the same approach that is used by https://github.com/mafintosh/csv-parser, which is very fast.

@mafintosh @maxogden I might need some of your help here

### CODE REVIEW

**Comment by mcollina** [comment]
Removed utf-8-validate, as not really needed.

**Comment by mcollina** [comment]
Wow, this is already 10% faster than the previous implementation.

The main difference here is that we depend on [bl](http://npm.im/bl).

Anybody sees a problem with that?

**Comment by mcollina** [comment]
@heroboy please test.

File: index.js [hunk 1]
```diff
@@ -17,30 +17,33 @@ IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
...
+  var needsSplit = chunk.toString('utf8').match(this.matcher)
+    , list
+    , remaining
+    , i
 
-  for (i = 0; i < list.length; i++) {
-    push(this, this.mapper(list[i]))
+  if (needsSplit) {
+    list = this._list.toString('utf8').split(this.matcher)
+    if (list) {
```
Comment (mafintosh) [inline index.js]: won't this always fire? [] is still thruthy.

Comment (mcollina) [inline index.js]: you are right, good spot. Removed.

**Comment by heroboy** [comment]
Thanks. It works.
I have two suggestions to improve performance.
1. Use https://nodejs.org/api/string_decoder.html to convert chunks to string, it will remember the last unused bytes. Then you don't need to store the buffer list.
2. add `{decodeStrings:false}` in `through` options,then if code like:
```
var s = split();
s.write("aaaaa");
```
the strings will not be converted to `Buffer`,and you convert it back again. But you need check the type of `chunk`.

**Comment by mcollina** [comment]
1. no you would need to store things, as that remember only the last part of a multibyte utf8 string, not the full line
2. possibly, but then I won't be able to use bl.

Why don't you send a PR, and we compare the perf?

**Comment by heroboy** [comment]
The second line is broken.
```
var s = split();
s.on("data", data=> { 
    console.log(data);
})

var str = "烫烫烫\r\n烫烫烫";
var buf = new Buffer(str, "utf8");
for (var i = 0; i < buf.length;i+=2)
{
    s.write(buf.slice(i, i + 2));
}
s.end();
```

**Comment by heroboy** [comment]
1. I don't know how to use git.
2. Version control system like git that need downloads all history , the usage experience is not good in China's network.
But I will try to make my version and compare the perf.

**Comment by mcollina** [comment]
Please try this one, it's based on StringDecoder, it should be good, and performance is equivalent as the current version.

**Comment by heroboy** [comment]
I think this one is right.
This is my bench result.
for small file "package.json"
```
benchSplit*10000: 1394.734ms
benchOldSplit*10000: 1327.251ms
benchThrough*10000: 1062.809ms
benchSplit*10000: 1125.691ms
benchOldSplit*10000: 1156.593ms
benchThrough*10000: 985.630ms
```
for large file(about 3MB)
```
benchSplit*100: 3900.208ms
benchOldSplit*100: 4079.545ms
benchThrough*100: 200.518ms
benchSplit*100: 3863.971ms
benchOldSplit*100: 3941.405ms
benchThrough*100: 195.780ms
```
So I think large file bench is better, it make big difference between `through()` and `split()`.

**Comment by mcollina** [comment]
what does `benchThrough`?

**Comment by heroboy** [comment]
Just replace `split()` with `require('through2')()`

**Comment by mcollina** [comment]
awesome, are you ok to merge then?

**Comment by heroboy** [comment]
Very ok.

**Comment by mcollina** [comment]
@mafintosh any other opinion on this one?

File: index.js [hunk 2]
```diff
@@ -17,24 +17,20 @@ IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
...
 
 var through = require('through2')
+var StringDecoder = require('string_decoder').StringDecoder
 
 function transform(chunk, enc, cb) {
-  var list = chunk.toString('utf8').split(this.matcher)
-    , remaining = list.pop()
+  this._last += this._decoder.write(chunk)
+
+  var list = this._last.toString('utf8').split(this.matcher)
```
Comment (mafintosh) [inline index.js]: toString() is not needed when we use a string decoder

**Comment by mafintosh** [comment]
@mcollina other than my minor nitpick comment i'm :+1:

**Comment by heroboy** [comment]
I think `flush` should be this:
```
function flush(cb) {
  var str = this._last + this._decoder.end();
  if (str )
    push(this, this.mapper(str))

  cb()
}
```

**Comment by yoshuawuyts** [comment]
@heroboy

> Version control system like git that need downloads all history

For future reference: `git` allows you to specify a `--depth` flag when cloning which greatly improves download speeds. Cheers! :sparkles:
```sh
$ git clone --depth 1 git@github.com:mcollina/split2.git
```

**Comment by mcollina** [comment]
@heroboy I've tried to have a test case where that `flush` is needed, but I can't come up with anything.

Also doing  `var str = this._last + this._decoder.end();` is going to return a broken utf8 string.

File: index.js (hunk 2 above)
Comment (mcollina) [inline index.js]: done :)

**Comment by heroboy** [comment]
I'm not very clear about the problem. If the input buffer is not a valid utf8 string, the `decoder.end()` will use last unused bytes to form an incorrect character, is this needed? And maybe future `StringDecoder` will cache bytes and emit strings later?

**Comment by mcollina** [comment]
`flush` is called after everything coming has been processed, so it means that an incomplete utf8 char has been written, and probably the stream has been cut. If there is anything left in the `StringDecoder`, probably it's better to emit an error, or just skip it.

@mafintosh any opinions on this one?

**Comment by heroboy** [comment]
`through({encoding:"utf8"})` this one doesn't emit an error.
```
var s = through({encoding:"utf8"});
s.on("data", data=> { 
    console.log("data",data);
})

var str = "烫烫烫\r\n烫烫烫";
var buf = new Buffer(str, "utf8");
for (var i = 0; i < buf.length-1;i+=1)
{
    s.write(buf.slice(i, i + 1));
}
s.end();
```

**Comment by mcollina** [comment]
@heroboy I've just pushed a test that reproduce the problem I'm speaking about:

https://github.com/mcollina/split2/blob/utf-8-support/test.js#L270-L285

This writes a truncated utf-8 char. If that char is not completed, it does not push anything down regarding that char.

**Comment by heroboy** [comment]
But in other node module, if the input utf8 buffer is not completed,it will output the last "incorrect" chars.
```
var fs = require("fs");
var buf = new Buffer("烫烫", "utf8");
fs.writeFileSync("broken.txt", buf.slice(0, buf.length - 1));
var s = fs.readFileSync("broken.txt", "utf8");
console.log(s);// output is "烫��"
```

**Comment by mcollina** [comment]
Done, let me know @heroboy.

**Comment by heroboy** [comment]
Great. I think there are no problems. And thank you very much make effort to fix this issue.

**Pull request merged**